File: Web/src/components/quotation/calc.ts
/**
 * Frontend port of backend/app/calc.py (clsDocComUtility.cs CalForm).
 * Used for live, instant amount recompute on every keystroke.
 * Must stay in sync with the Python version.
 */

import type { QuotationLine, ComputeResult } from '@/lib/api'

const GRP_STOCK      = new Set(['Stock', 'Non_Stock', 'Service', 'Assembly'])
const GRP_CHARGES    = new Set(['Charges'])
const GRP_DISCOUNT   = new Set(['Discount'])
const GRP_TOTAL      = new Set(['Sub Total', 'Total', 'CF Total'])
const GRP_STRUCTURAL = new Set(['Header', 'Remark'])

function rnd4(n: number): number {
  return Math.round(n * 10000) / 10000
}

export function computeLines(
  lines: QuotationLine[],
  currencyRate = 1,
  docDiscountRate = 0,
  taxRate = 9,
): ComputeResult {
  const sorted = [...lines].sort((a, b) => a.line_no - b.line_no)

  let totalPrevious = 0
  let totalSt = 0
  let totalCf = 0
  let detTotalAmt = 0
  let taxableTotal = 0
  let resetTotal = false

  const computedLines: ComputeResult['lines'] = []

  for (const row of sorted) {
    const lt = row.line_type
    let price = row.price ?? 0
    const qty = row.qty ?? 0
    const taxable = row.taxable ?? true
    const rowTaxRate = row.tax_rate ?? 0

    let amt = 0
    let taxAmt = 0
    let taxAmtLocal = 0

    if (GRP_STOCK.has(lt)) {
      amt = rnd4(row.amount ?? 0)
      if (taxable && rowTaxRate > 0) taxAmt = rnd4(amt * rowTaxRate / 100)
      taxAmtLocal = rnd4(taxAmt * currencyRate)
      detTotalAmt += amt
      if (taxable) taxableTotal += amt
      totalCf += amt
      if (resetTotal) { totalSt = amt; resetTotal = false } else { totalSt += amt }

    } else if (GRP_CHARGES.has(lt)) {
      amt = rnd4(row.amount ?? 0)
      if (taxable && rowTaxRate > 0) taxAmt = rnd4(amt * rowTaxRate / 100)
      taxAmtLocal = rnd4(taxAmt * currencyRate)
      detTotalAmt += amt
      if (taxable) taxableTotal += amt
      totalCf += amt
      if (resetTotal) { totalSt = amt; resetTotal = false } else { totalSt += amt }

    } else if (GRP_DISCOUNT.has(lt)) {
      amt = rnd4(row.amount ?? 0)
      detTotalAmt += amt
      totalCf += amt
      if (resetTotal) { totalSt = amt; resetTotal = false } else { totalSt += amt }

    } else if (GRP_TOTAL.has(lt)) {
      if (lt === 'Sub Total') amt = totalSt
      else if (lt === 'CF Total') amt = totalCf
      else amt = totalPrevious + totalSt   // Total

      totalPrevious = amt
      resetTotal = true
      price = 0
      taxAmt = 0
      taxAmtLocal = 0

    } else {
      // Header / Remark – zero contribution
      amt = 0; price = 0; taxAmt = 0; taxAmtLocal = 0
    }

    computedLines.push({
      line_no: row.line_no,
      price,
      amount: amt,
      tax_amt: taxAmt,
      tax_amt_local: taxAmtLocal,
    })
  }

  const subTotal       = rnd4(detTotalAmt)
  const discountAmt    = rnd4(subTotal * docDiscountRate / 100)
  const afterDis       = rnd4(subTotal - discountAmt)
  const taxableAfterDis = rnd4(taxableTotal - rnd4(taxableTotal * docDiscountRate / 100))
  const taxTotal       = rnd4(taxableAfterDis * taxRate / 100)
  const grandTotal     = rnd4(afterDis + taxTotal)
  const homeSubTotal   = rnd4(subTotal  * currencyRate)
  const homeTaxTotal   = rnd4(taxTotal  * currencyRate)
  const homeTotal      = rnd4(afterDis  * currencyRate) + homeTaxTotal

  return {
    lines: computedLines,
    sub_total:       subTotal,
    discount_amt:    discountAmt,
    total_after_dis: afterDis,
    tax_total:       taxTotal,
    grand_total:     grandTotal,
    home_sub_total:  homeSubTotal,
    home_tax_total:  homeTaxTotal,
    home_total:      homeTotal,
  }
}

/** Types where price entry is blocked in the UI */
export const PRICE_FORBIDDEN = new Set([...GRP_TOTAL, ...GRP_STRUCTURAL])
/** Types where qty entry is blocked in the UI */
export const QTY_FORBIDDEN   = new Set([...GRP_CHARGES, ...GRP_TOTAL, ...GRP_STRUCTURAL])
/** Types that don't require an item selection */
export const ITEM_OPTIONAL   = new Set([...GRP_TOTAL, ...GRP_STRUCTURAL])

export const ALL_LINE_TYPES = [
  'Header', 'Charges', 'Stock', 'Assembly', 'Non_Stock', 'Service', 'Discount', 'Remark',
  'Sub Total', 'Total', 'CF Total',
]

File: backend/app/calc.py
"""
Quotation computation engine – ported from clsDocComUtility.cs CalForm.

Single source of truth for amount rollup. Also used by the frontend (calc.ts).
"""
from __future__ import annotations
from decimal import Decimal, ROUND_HALF_UP
from typing import Any

from app.enums import (
    GRP_STOCK, GRP_CHARGES, GRP_DISCOUNT,
    GRP_TOTAL, GRP_STRUCTURAL,
    LINE_TYPE_SUB_TOTAL, LINE_TYPE_TOTAL, LINE_TYPE_CF_TOTAL,
)

_ZERO = Decimal("0")
_CENT = Decimal("0.0001")  # 4 dp rounding for money


def _rnd(v: Decimal) -> Decimal:
    return v.quantize(_CENT, rounding=ROUND_HALF_UP)


def compute(
    lines: list[dict[str, Any]],
    currency_rate: Decimal = Decimal("1"),
    doc_discount_rate: Decimal = _ZERO,
    tax_rate: Decimal = Decimal("9"),
) -> dict[str, Any]:
    """
    Iterate lines (sorted by line_no) and compute per-line amounts + document totals.

    Returns a dict:
      {
        "lines": [ {line_no, amount, tax_amt, tax_amt_local, price}, ... ],
        "sub_total", "discount_amt", "total_after_dis",
        "tax_total", "grand_total",
        "home_sub_total", "home_tax_total", "home_total"
      }

    Mirrors CalForm logic (clsDocComUtility.cs ~550-1843):
      - Stock/Non_Stock/Service/Assembly/Charges/Discount: amount = row.amount (user-entered)
      - Sub Total row: amount = total_st (sum since last reset)
      - Total row: amount = total_previous + total_st
      - CF Total row: amount = total_cf (running cumulative)
      - Header/Remark: contribute 0 to det_total_amt
    """
    sorted_lines = sorted(lines, key=lambda r: r.get("line_no", 0))

    total_previous = _ZERO   # last Total row value
    total_st       = _ZERO   # running sub-total since last reset
    total_cf       = _ZERO   # cumulative carry-forward
    det_total_amt  = _ZERO   # feeds doc sub_total (excludes Header/Total rows)
    taxable_total  = _ZERO
    reset_total    = False

    computed_lines: list[dict[str, Any]] = []

    for row in sorted_lines:
        lt = row.get("line_type", "")
        price = Decimal(str(row.get("price", 0)))
        qty   = Decimal(str(row.get("qty", 0)))
        taxable = bool(row.get("taxable", True))
        row_tax_rate = Decimal(str(row.get("tax_rate", 0)))
        line_no = row.get("line_no", 0)

        amt = _ZERO
        tax_amt = _ZERO
        tax_amt_local = _ZERO

        if lt in GRP_STOCK:
            amt = _rnd(Decimal(str(row.get("amount", 0))))
            if taxable and row_tax_rate > 0:
                tax_amt = _rnd(amt * row_tax_rate / 100)
            tax_amt_local = _rnd(tax_amt * currency_rate)
            det_total_amt += amt
            if taxable:
                taxable_total += amt
            total_cf += amt
            if reset_total:
                total_st = amt
                reset_total = False
            else:
                total_st += amt

        elif lt in GRP_CHARGES:
            amt = _rnd(Decimal(str(row.get("amount", 0))))
            if taxable and row_tax_rate > 0:
                tax_amt = _rnd(amt * row_tax_rate / 100)
            tax_amt_local = _rnd(tax_amt * currency_rate)
            det_total_amt += amt
            if taxable:
                taxable_total += amt
            total_cf += amt
            if reset_total:
                total_st = amt
                reset_total = False
            else:
                total_st += amt

        elif lt in GRP_DISCOUNT:
            amt = _rnd(Decimal(str(row.get("amount", 0))))
            det_total_amt += amt
            total_cf += amt
            if reset_total:
                total_st = amt
                reset_total = False
            else:
                total_st += amt

        elif lt in GRP_TOTAL:
            # Total rollup rows (clsDocComUtility.cs:1507-1590)
            if lt == LINE_TYPE_SUB_TOTAL:
                amt = total_st
            elif lt == LINE_TYPE_CF_TOTAL:
                amt = total_cf
            else:  # Total
                amt = total_previous + total_st

            total_previous = amt
            reset_total = True
            # price/tax forced to zero for these rows
            price = _ZERO
            tax_amt = _ZERO
            tax_amt_local = _ZERO

        else:
            # Header / Remark (clsDocComUtility.cs:1595-1629)
            # Does not contribute to det_total_amt
            amt = _ZERO
            price = _ZERO
            tax_amt = _ZERO
            tax_amt_local = _ZERO

        computed_lines.append({
            "line_no": line_no,
            "price": float(price),
            "amount": float(amt),
            "tax_amt": float(tax_amt),
            "tax_amt_local": float(tax_amt_local),
        })

    # Document totals (clsDocComUtility.cs:1784-1843)
    sub_total    = _rnd(det_total_amt)
    discount_amt = _rnd(sub_total * doc_discount_rate / 100)
    after_dis    = sub_total - discount_amt
    taxable_after_dis = _rnd(taxable_total - _rnd(taxable_total * doc_discount_rate / 100))
    tax_total    = _rnd(taxable_after_dis * tax_rate / 100)
    grand_total  = after_dis + tax_total

    home_sub     = _rnd(sub_total  * currency_rate)
    home_tax     = _rnd(tax_total  * currency_rate)
    home_total   = _rnd(after_dis  * currency_rate) + home_tax

    return {
        "lines": computed_lines,
        "sub_total":       float(sub_total),
        "discount_amt":    float(discount_amt),
        "total_after_dis": float(after_dis),
        "tax_total":       float(tax_total),
        "grand_total":     float(grand_total),
        "home_sub_total":  float(home_sub),
        "home_tax_total":  float(home_tax),
        "home_total":      float(home_total),
    }
